src/include/miopen/env.hpp

```diff
@@ -31,131 +31,174 @@
+// Returns 0 if environment variable is in wrong format (strtoull fails to parse the string).
+template <>
+struct ParseEnvVal<uint64_t>
 {
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
+    static uint64_t go(const char* vp) { return std::strtoull(vp, nullptr, 0); }
+};
+
+template <>
+struct ParseEnvVal<std::string>
+{
+    static std::string go(const char* vp) { return std::string{vp}; }
+};
+
+template <typename T>
+struct EnvVar
+{
+private:
+    T value{};
+    bool is_unset = true;
+
+public:
+    const T& GetValue() const { return value; }
+
+    bool IsUnset() const { return is_unset; }
+
+    void UpdateValue(const T& val)
     {
-        return false;
+        is_unset = false;
+        value    = val;
     }
-    else
+
+    explicit EnvVar(const char* const name, const T& def_val)
     {
-        std::string value_env_str = value_env_p;
-        for(auto& c : value_env_str)
+        // NOLINTNEXTLINE (concurrency-mt-unsafe)
+        const char* vp = std::getenv(name);
+        if(vp != nullptr) // a value was provided
         {
-            if(std::isalpha(c) != 0)
-            {
-                c = std::tolower(static_cast<unsigned char>(c));
-            }
+            is_unset = false;
+            value    = ParseEnvVal<T>::go(vp);
+        }
+        else // no value provided, use default value
+        {
+            value = def_val;
         }
-        return (std::strcmp(value_env_str.c_str(), "enable") == 0 ||
-                std::strcmp(value_env_str.c_str(), "enabled") == 0 ||
-                std::strcmp(value_env_str.c_str(), "1") == 0 ||
-                std::strcmp(value_env_str.c_str(), "yes") == 0 ||
-                std::strcmp(value_env_str.c_str(), "on") == 0 ||
-                std::strcmp(value_env_str.c_str(), "true") == 0);
     }
-}
+};
 
-// Return 0 if env is enabled else convert environment var to an int.
-// Supports hexadecimal with leading 0x or decimal
-inline uint64_t EnvvarValue(const char* name, uint64_t fallback = 0)
-{
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
-    {
-        return fallback;
-    }
-    else
-    {
-        return strtoull(value_env_p, nullptr, 0);
+} // end namespace internal
+
+// static inside function hides the variable and provides
+// thread-safety/locking
+// declare in global namespace
+#define MIOPEN_DECLARE_ENV_VAR(name, type, default_val)                    \
+    namespace miopen::env {                                                \
+    struct name                                                            \
+    {                                                                      \
+        static_assert(std::is_same_v<name, ::miopen::env::name>,           \
+                      "must be in miopen::env and must be unique");        \
```

Comment:
[R]
```suggestion
                      "This macro must be used in miopen::env namespace");        \
```
Uniqueness is not required - the same declaration may appear in several translation units.

Reply:
updated assert failure string

Reply:
[resolved] and yeah, my suggestion was incorrect